README.md
_PageObject_ is an abstraction meaning part of a web page maintating its specific logic over its components (DOM elements and other page objects when necessary).

While creating rich web applications we usually single out page blocks. We refer to those blocks as objects when we need to repeat them somehow here or there. This is where the utility comes to hand. **PageObject.js** would help you if for such cases you used to render client-side templates and then revive their rendered results.

**PageObject.js** is a jQuery utility and there are (as of v0.14) actually only two functions:

- `$.turnToPageObject`and
- `$.turnToPageObject.configure`.

---

## $.turnToPageObject( _[object]_ target, _[object]_ options )

The idea is that `target` should get `target.DOM` namespace filled with HTML DOM elements according to `options.selectors` taken from `options.template` rendered  with `options.context`.

```javascript
function LoginForm(successCallback) {
  var lf = this;
  $.turnToPageObject(lf, {
    template: '<form>'+
              '<h1><%= caption %></h1>'+
              '<input type="text" name="email" />'+
              '<input type="password" name="password" />'+
              '<input type="submit" name="submit" />'
              '</form',
    context: {
      caption: I18n.t("Please enter your credentials below")
    },
    selectors: {
      form: 'form',
      email: 'input[name=email]',
      password: ':password'
    }
  });

  $(lf.DOM.form).submit(function (e) {
    e.preventDefault();
    var email = $(lf.DOM.email).val(),
      password = $(lf.DOM.password).val(),
      ok = authenticate(email, password);
      if (!ok) {
        alert("No way, buddy!");
      } else {
        success();
      }
  });
}

var loginForm = new LoginForm(function () {
  // render some other view
});

$('body').append(loginForm.DOM.container);
```

### $.turnToPageObject options

#### Container related

If you ommit `container` _DOM element_ it will be created for you from `containerElement` which is **DIV** by default.

If you provide `containerClass` _string_ option, your `target.DOM.container` will get that class.

If you set `hide` to `true` option, `target.DOM.container` will be hidden. This has proved to be useful.

#### Template related

if you provide `template` option, the rendered template will be placed inside `target.DOM.container`.

If you omit `context` option, template will be rendered with empty context `{}`.

`template` can be either a function or a string.

- Function `template` takes single argument, a `context` object, and returns string. This is made for integration with [Jammit JST](http://documentcloud.github.io/jammit/#jst).
- String templates are rendered with `templateEngine` which takes two arguments, `template` string and `context` object. `templateEngine` function also returns string.

`templateEngine` is being automatically configured to use `_.template` if [Underscore](http://documentcloud.github.io/underscore) is present.

If there's no _Underscore_ in your project, you'll need to configure `templateEngine`.

This is how to set defaults:
```javascript
$.turnToPageObject.configure({
  templateEngine: window.tmpl, // http://ejohn.org/blog/javascript-micro-templating
  containerElement: 'strong'
});
```

#### Selectors

If you provide `selectors` _object_ option, HTML DOM elements will be found within `target.DOM.container` and placed in `target.DOM` under corresponding keys. If `template` option was specified, `selectors` will be searching after it is rendered into the container.

If a selector fails to find any element it will throw an exception. If it finds multiple elements it will shout too. If you do really need to find multiple selectors and store them in array under the corresponding key, just prepend `[]` (square brackets) to a selector, e.g. '[] ul > li'.

If you need to collect some elements into an object having their keys determined from HTML DOM elements found, like in the [Calculator](http://j.mp/Z1roQS) demo, write selector as a two elements array `[ realSelector, functionToDetermineElementsKey ]`.

Namespaces in `selectors` result in same namespaces in `target.DOM`.

---

Please check out [the demo](http://j.mp/Z1roQS), [the source code](https://github.com/IronGroove/PageObject.js) and tests to learn more about how things work.


CHANGELOG
v0.16.1
  Bugfix: Container is no more created when there is already one.

v0.16
  Question mark selector feature.
  Shout for [] and ? features when selector is an array.
  `new $.PageObject(opts)` feature
  Namespace option feature.

v0.15.1
  Fix for

v0.15
  priv.isElement => $.isElement
  string errors => new Error()
  priv.extractParts => $.extractParts
  $.isPageObject(po)
  $(po.DOM.container).data('pageObject', po)

v0.14
  One day…


demo/tmpl.js
// Simple JavaScript Templating
// John Resig - http://ejohn.org/ - MIT Licensed
(function(){
  var cache = {};

  this.tmpl = function tmpl(str, data){
    // Figure out if we're getting a template, or if we need to
    // load the template - and be sure to cache the result.
    var fn = !/\W/.test(str) ?
      cache[str] = cache[str] ||
        tmpl(document.getElementById(str).innerHTML) :

      // Generate a reusable function that will serve as a template
      // generator (and which will be cached).
      new Function("obj",
        "var p=[],print=function(){p.push.apply(p,arguments);};" +

        // Introduce the data as local variables using with(){}
        "with(obj){p.push('" +

        // Convert the template into pure JavaScript
//          str
//            .replace(/[\r\t\n]/g, " ")
//            .split("<%").join("\t")
//            .replace(/((^|%>)[^\t]*)'/g, "$1\r")
//            .replace(/\t=(.*?)%>/g, "',$1,'")
//            .split("\t").join("');")
//            .split("%>").join("p.push('")
//            .split("\r").join("\\'")
//          + "');}return p.join('');");

          // Neil Donewar fix for the multiple quotes issue.
          // http://ejohn.org/blog/javascript-micro-templating/#comment-321850
          str.replace(/[\r\t\n]/g, " ")
             .replace(/'(?=[^%]*%>)/g,"\t")
             .split("'").join("\\'")
             .split("\t").join("'")
             .replace(/<%=(.+?)%>/g, "',$1,'")
             .split("<%").join("');")
             .split("%>").join("p.push('")
             + "');}return p.join('');");

    // Provide some basic currying to the user
    return data ? fn( data ) : fn;
  };
})();


demo/Calculator.js
(function ($) {

  function Calculator() {
    var history = [],
      calc = this,
      $led;

    $.turnToPageObject(calc, {
      template: $('#tmplCalculator').html(),
      containerClass: 'calc',
      context: {
        caption: "Demo Calculator",
      },
      selectors: {
        buttons: [ ':button', Calculator.getButtonName ],
        led: 'p'
      }
    });

    $led = $(calc.DOM.led);

    function remember() {
      var current = $led.html();
      if (current == history[history.length-1]) return;
      history.push( $led.html() );
    }

    for (var name in calc.DOM.buttons)
      $(calc.DOM.buttons[name]).click(function () {
        var name = Calculator.getButtonName(this);
        if (name == parseInt(name)) {
          remember();
          $led.html( $led.html() + ' ' + name );
        }
        else switch (name) {
          case "cancel":
            history = [];
            $led.html('');
            break;
          case "undo":
            $led.html(history.pop());
            break;
          case "plus":
            remember();
            $led.html( $led.html() + ' + ' );
            break;
          case "minus":
            remember();
            $led.html( $led.html() + ' - ' );
            break;
          case "multiply":
            remember();
            $led.html( $led.html() + ' * ' );
            break;
          case "divide":
            remember();
            $led.html( $led.html() + ' / ' );
            break;
          case "equals":
            var result,
              statement = $led.html().replace(/ /g, '');
            try {
              result = eval(statement);
            } catch(e) {}
            if (result !== undefined) {
              remember();
              $led.html( (result+"").replace(/./g, "$& ") );
            }
            break;
          case "dot":
            remember();
            $led.html( $led.html() + ' .' );
            break;
        }
      });

    return calc;
  }

  Calculator.getButtonName = function (el) {
    return $(el).attr('class').replace('calc-','');
  }

  window.Calculator = Calculator;

})(jQuery);


releases/PageObject-0.14.js
/*global jQuery: false */
/*jslint browser: true, indent: 2, white: true, nomen: true */

/*
 * PageObject v0.14
 *
 * Copyright 2011, Mykhaylo Gavrylyuk, https://github.com/IronGroove
 * Licensed under the MIT license
 *
 * Date: Thu May 16 23:21:14 EEST 2013
 */

(function (window, $) {
  "use strict";

  var priv = {};

  priv.defaultOptions = {
    container: undefined,
    containerClass: undefined,
    containerElement: 'div',
    templateEngine: undefined,
    template: undefined,
    context: {},
    selectors: {},
    hide: false
  };

  // Use underscore microtemplating function as default if using underscore.
  if (window._ && $.isFunction(window._.template)) {
    priv.defaultOptions.templateEngine = window._.template;
  }

  // Extract jQuery DOM parts from a container using a map of selectors.
  priv.extractParts = function (sourceContainer, selectors) {
    var domParts = {};

    if (!priv.isElement(sourceContainer)) {
      throw "POE10: arguments should be 1st — HTML DOM container and 2nd — selectors map";
    }

    if (!$.isPlainObject(selectors)) {
      selectors = {};
    }


    $.each(selectors, function (name, selector) {
      var found,
        findMultiple = false;

      if (!/^[A-z\.]+$/.test(name)) {
        throw "POE12: incorrect selector name `" + name + "`";
      }

      if ($.isArray(selector) &&
        selector.length === 2 &&
        $.type(selector[0]) === 'string' &&
        $.isFunction(selector[1])) {

        domParts[name] = {};
        $(sourceContainer).find(selector[0]).each(function () {
          var id = selector[1].call(this, this);
          if (id) {
            if (domParts[name][id]) {
              throw "POE13: duplicate identifier `" + id + "` in DOM part namespace `"+ name +"`";
            }
            domParts[name][id] = this;
          }
        });
      }

      else if ($.isPlainObject(selector)) {
        domParts[name] = priv.extractParts(sourceContainer, selector);
      }

      else if (typeof selector === 'string') {
        findMultiple = selector.indexOf('[]') === 0;
        if (findMultiple) {
          selector = selector.replace('[]', '');
        }
        found = $(sourceContainer).find(selector);
        if (found.length === 0) {
          throw "POE14: DOM parts weren't found for selector `" + name + "`";
        } else if (found.length > 1) {
          if (findMultiple) {
            found = Array.prototype.slice.call(found);
          } else {
            throw "POE15: multiple DOM parts found for selector `" + name + "`";
          }
        } else {
          found = found[0];
        }
        domParts[name] = found;
      }

      else {
        throw "POE16: invalid selector value";
      }

    });

    return domParts;
  };

  // http://stackoverflow.com/a/384380/133257
  priv.isElement = function (o) {
    return  (
      typeof window.HTMLElement === "object" ? o instanceof window.HTMLElement : // DOM2
      typeof o === "object" && o !== null && o.nodeType === 1 && typeof o.nodeName === "string"
    );
  };

  $.turnToPageObject = function (target, options) {
    var opts, rendered, domParts;


    //   CHECK IF 1ST ARGUMENT IS OBJECT
    //   -------------------------------

    if ($.type(target) !== 'object') {
      throw "POE01: invalid target (1st argument should be an object)!";
    }

    if (!$.isPlainObject(options)) {
      options = {};
    }

    opts = $.extend({}, priv.defaultOptions, options);


    //   CREATE DOM NAMESPACE IF THERE'S NO SUCH
    //   ---------------------------------------

    if (!$.isPlainObject(target.DOM)) {
      target.DOM = {};
    }


    //   PREPARE CONTAINER
    //   -----------------

    if (opts.container !== undefined) {
      if (!priv.isElement(opts.container)) {
        throw "POE02: container is invalid!";
      }
      target.DOM.container = opts.container;
    } else {
      target.DOM.container = window.document.createElement(opts.containerElement);
    }


    //   SET CONTAINER CLASS WHEN NEEDED
    //   -------------------------------

    if (opts.containerClass) {
      $(target.DOM.container).addClass(opts.containerClass);
    }



    //   RENDER TEMPLATE
    //   ---------------

    if (opts.template !== undefined) {

      if (typeof opts.template === 'string') {
        if (!$.isFunction(opts.templateEngine)) {
          throw "POE03: templateEngine not configured";
        }
      } else if (typeof opts.template !== 'function' ) {
        throw "POE04: invalid template";
      }


      if (!$.isPlainObject(opts.context)) {
        opts.context = {};
      }

      try {
        if (typeof opts.template === 'string') {
          rendered = opts.templateEngine(opts.template, opts.context);
        } else {
          rendered = opts.template(opts.context);
        }
      } catch (e) {
        throw "POE20: template error: " + e;
      }

      $(target.DOM.container).html(rendered);
    }


    //   EXTRACTING DOM PARTS
    //   --------------------

    if (!$.isPlainObject(opts.selectors)) {
      opts.selectors = {};
    }

    domParts = priv.extractParts(target.DOM.container, opts.selectors);
    $.extend(true, target.DOM, domParts);


    //   HIDE THE CONTAINER IF IT IS NECESSARY
    //   -------------------------------------

    if (opts.hide) {
      $(target.DOM.container).hide();
    }

    return target;
  };

  $.turnToPageObject.configure = function (options) {
    if ($.isPlainObject(options)) {
      $.extend(priv.defaultOptions, options);
    }
  };

  // Expose private things for testing.
  // Don't worry, following lines are not included
  // in minimized production version.
  if (window.PAGE_OBJECT_JS_TEST_MODE) {
    $.turnToPageObject.priv = priv;
  }

}(window, jQuery));


releases/PageObject-0.15.js
/*global jQuery: false */
/*jslint browser: true, indent: 2, white: true, nomen: true */

/*
 * PageObject v0.15
 *
 * Copyright 2011, Mykhaylo Gavrylyuk, https://github.com/IronGroove
 * Licensed under the MIT license
 *
 * Date: Fri May 31 00:15:25 EEST 2013
 */

(function (window, $) {
  "use strict";

  var priv = {}, POE;

  function PageObjectError(code, message) {
    this.name = "PageObjectError";
    this.code = code < 10 ? "0"+code : code;
    this.message = message;
  }
  PageObjectError.prototype = new Error();
  PageObjectError.prototype.constructor = new PageObjectError;
  PageObjectError.prototype.toString = function () {
    return "POE" + this.code + ": " + this.message;
  }

  POE = PageObjectError;

  priv.defaultOptions = {
    container: undefined,
    containerClass: undefined,
    containerElement: 'div',
    templateEngine: undefined,
    template: undefined,
    context: {},
    selectors: {},
    hide: false
  };

  // Use underscore microtemplating function as default if using underscore.
  if (window._ && $.isFunction(window._.template)) {
    priv.defaultOptions.templateEngine = window._.template;
  }

  // Extract jQuery DOM parts from a container using a map of selectors.
  $.extractParts = function (sourceContainer, selectors) {
    var domParts = {};

    if (!$.isElement(sourceContainer)) {
      throw new POE(10, "arguments should be: 1st — HTML DOM container, 2nd — selectors map");
    }

    if (!$.isPlainObject(selectors)) {
      selectors = {};
    }

    $.each(selectors, function (name, selector) {
      var found,
        findMultiple = false;

      if (!/^[A-z\.]+$/.test(name)) {
        throw new POE(11, "incorrect selector name `" + name + "`");
      }

      if ($.isArray(selector) &&
        selector.length === 2 &&
        $.type(selector[0]) === 'string' &&
        $.isFunction(selector[1])) {

        domParts[name] = {};
        $(sourceContainer).find(selector[0]).each(function () {
          var id = selector[1].call(this, this);
          if (id) {
            if (domParts[name][id]) {
              throw new POE(12, "duplicate identifier `" + id + "` in DOM part namespace `"+ name +"`");
            }
            domParts[name][id] = this;
          }
        });
      }

      else if ($.isPlainObject(selector)) {
        domParts[name] = $.extractParts(sourceContainer, selector);
      }

      else if (typeof selector === 'string') {
        findMultiple = selector.indexOf('[]') === 0;
        if (findMultiple) {
          selector = selector.replace('[]', '');
        }
        found = $(sourceContainer).find(selector);
        if (found.length === 0) {
          throw new POE(13, "DOM parts weren't found for selector `" + name + "`");
        } else if (found.length > 1) {
          if (findMultiple) {
            found = Array.prototype.slice.call(found);
          } else {
            throw new POE(14, "multiple DOM parts found for selector `" + name + "`");
          }
        } else {
          found = found[0];
        }
        domParts[name] = found;
      }

      else {
        throw new POE(15, "invalid selector value");
      }

    });

    return domParts;
  };

  // http://stackoverflow.com/a/384380/133257
  $.isElement = function (o) {
    return  (
      typeof window.HTMLElement === "object" ? o instanceof window.HTMLElement : // DOM2
      typeof o === "object" && o !== null && o.nodeType === 1 && typeof o.nodeName === "string"
    );
  };

  $.isPageObject = function (po) {
    return po && po.DOM && $.isPlainObject(po.DOM) && $.isElement(po.DOM.container);
  }

  $.turnToPageObject = function (target, options) {
    var opts, rendered, domParts;


    //   CHECK IF 1ST ARGUMENT IS OBJECT
    //   -------------------------------

    if ($.type(target) !== 'object') {
      throw new POE(1, "invalid target (1st argument should be an object)!");
    }

    if (!$.isPlainObject(options)) {
      options = {};
    }

    opts = $.extend({}, priv.defaultOptions, options);


    //   CREATE DOM NAMESPACE IF THERE'S NO SUCH
    //   ---------------------------------------

    if (!$.isPlainObject(target.DOM)) {
      target.DOM = {};
    }


    //   PREPARE CONTAINER
    //   -----------------

    if (opts.container !== undefined) {
      if (!$.isElement(opts.container)) {
        throw new POE(2, "container is invalid!");
      }
      target.DOM.container = opts.container;
    } else {
      target.DOM.container = window.document.createElement(opts.containerElement);
    }


    //   SET CONTAINER CLASS WHEN NEEDED
    //   -------------------------------

    if (opts.containerClass) {
      $(target.DOM.container).addClass(opts.containerClass);
    }


    //   ESTABLISH REFERENCE
    //   -------------------

    $(target.DOM.container).data('pageObject', target);


    //   RENDER TEMPLATE
    //   ---------------

    if (opts.template !== undefined) {

      if (typeof opts.template === 'string') {
        if (!$.isFunction(opts.templateEngine)) {
          throw new POE(3, "templateEngine not configured");
        }
      } else if (typeof opts.template !== 'function' ) {
        throw new POE(4, "invalid template");
      }


      if (!$.isPlainObject(opts.context)) {
        opts.context = {};
      }

      try {
        if (typeof opts.template === 'string') {
          rendered = opts.templateEngine(opts.template, opts.context);
        } else {
          rendered = opts.template(opts.context);
        }
      } catch (e) {
        throw new POE(5, "template error: " + e);
      }

      $(target.DOM.container).html(rendered);
    }


    //   EXTRACTING DOM PARTS
    //   --------------------

    if (!$.isPlainObject(opts.selectors)) {
      opts.selectors = {};
    }

    domParts = $.extractParts(target.DOM.container, opts.selectors);
    $.extend(true, target.DOM, domParts);


    //   HIDE THE CONTAINER IF IT IS NECESSARY
    //   -------------------------------------

    if (opts.hide) {
      $(target.DOM.container).hide();
    }

    return target;
  };

  $.turnToPageObject.configure = function (options) {
    if ($.isPlainObject(options)) {
      $.extend(priv.defaultOptions, options);
    }
  };

  // Expose private things for testing.
  // Don't worry, following lines are not included
  // in minimized production version.
  if (window.PAGE_OBJECT_JS_TEST_MODE) {
    $.turnToPageObject.priv = priv;
  }

}(window, jQuery));


releases/PageObject-0.15.1.js
/*global jQuery: false */
/*jslint browser: true, indent: 2, white: true, nomen: true */

/*
 * PageObject v0.15.1
 *
 * Copyright 2011-2013, Mykhaylo Gavrylyuk, https://github.com/IronGroove
 * Licensed under the MIT license
 *
 * Date: Fri Jun 07 04:26:17 EEST 2013
 */

(function (window, $) {
  "use strict";

  var priv = {}, POE;

  function PageObjectError(code, message) {
    this.name = "PageObjectError";
    this.code = code < 10 ? "0"+code : code;
    this.message = message;
  }
  function Mediator(){}
  Mediator.prototype = Error.prototype;
  PageObjectError.prototype = new Mediator();
  PageObjectError.prototype.constructor = new PageObjectError;
  PageObjectError.prototype.toString = function () {
    return "PageObjectError #" + this.code + ": " + this.message;
  }

  POE = PageObjectError;

  priv.defaultOptions = {
    container: undefined,
    containerClass: undefined,
    containerElement: 'div',
    templateEngine: undefined,
    template: undefined,
    context: {},
    selectors: {},
    hide: false
  };

  // Use underscore microtemplating function as default if using underscore.
  if (window._ && $.isFunction(window._.template)) {
    priv.defaultOptions.templateEngine = window._.template;
  }

  // Extract jQuery DOM parts from a container using a map of selectors.
  $.extractParts = function (sourceContainer, selectors) {
    var domParts = {};

    if (!$.isElement(sourceContainer)) {
      throw new POE(10, "arguments should be: 1st — HTML DOM container, 2nd — selectors map");
    }

    if (!$.isPlainObject(selectors)) {
      selectors = {};
    }

    $.each(selectors, function (name, selector) {
      var found,
        findMultiple = false;

      if (!/^[A-z\.]+$/.test(name)) {
        throw new POE(11, "incorrect selector name `" + name + "`");
      }

      if ($.isArray(selector) &&
        selector.length === 2 &&
        $.type(selector[0]) === 'string' &&
        $.isFunction(selector[1])) {

        domParts[name] = {};
        $(sourceContainer).find(selector[0]).each(function () {
          var id = selector[1].call(this, this);
          if (id) {
            if (domParts[name][id]) {
              throw new POE(12, "duplicate identifier `" + id + "` in DOM part namespace `"+ name +"`");
            }
            domParts[name][id] = this;
          }
        });
      }

      else if ($.isPlainObject(selector)) {
        domParts[name] = $.extractParts(sourceContainer, selector);
      }

      else if (typeof selector === 'string') {
        findMultiple = selector.indexOf('[]') === 0;
        if (findMultiple) {
          selector = selector.replace('[]', '');
        }
        found = $(sourceContainer).find(selector);
        if (found.length === 0) {
          throw new POE(13, "DOM parts weren't found for selector `" + name + "`");
        } else if (found.length > 1) {
          if (findMultiple) {
            found = Array.prototype.slice.call(found);
          } else {
            throw new POE(14, "multiple DOM parts found for selector `" + name + "`");
          }
        } else {
          found = found[0];
        }
        domParts[name] = found;
      }

      else {
        throw new POE(15, "invalid selector value");
      }

    });

    return domParts;
  };

  // http://stackoverflow.com/a/384380/133257
  $.isElement = function (o) {
    return  (
      typeof window.HTMLElement === "object" ? o instanceof window.HTMLElement : // DOM2
      typeof o === "object" && o !== null && o.nodeType === 1 && typeof o.nodeName === "string"
    );
  };

  $.isPageObject = function (po) {
    return po && po.DOM && $.isPlainObject(po.DOM) && $.isElement(po.DOM.container);
  }

  $.turnToPageObject = function (target, options) {
    var opts, rendered, domParts;


    //   CHECK IF 1ST ARGUMENT IS OBJECT
    //   -------------------------------

    if ($.type(target) !== 'object') {
      throw new POE(1, "invalid target (1st argument should be an object)!");
    }

    if (!$.isPlainObject(options)) {
      options = {};
    }

    opts = $.extend({}, priv.defaultOptions, options);


    //   CREATE DOM NAMESPACE IF THERE'S NO SUCH
    //   ---------------------------------------

    if (!$.isPlainObject(target.DOM)) {
      target.DOM = {};
    }


    //   PREPARE CONTAINER
    //   -----------------

    if (opts.container !== undefined) {
      if (!$.isElement(opts.container)) {
        throw new POE(2, "container is invalid!");
      }
      target.DOM.container = opts.container;
    } else {
      target.DOM.container = window.document.createElement(opts.containerElement);
    }


    //   SET CONTAINER CLASS WHEN NEEDED
    //   -------------------------------

    if (opts.containerClass) {
      $(target.DOM.container).addClass(opts.containerClass);
    }


    //   ESTABLISH REFERENCE
    //   -------------------

    $(target.DOM.container).data('pageObject', target);


    //   RENDER TEMPLATE
    //   ---------------

    if (opts.template !== undefined) {

      if (typeof opts.template === 'string') {
        if (!$.isFunction(opts.templateEngine)) {
          throw new POE(3, "templateEngine not configured");
        }
      } else if (typeof opts.template !== 'function' ) {
        throw new POE(4, "invalid template");
      }


      if (!$.isPlainObject(opts.context)) {
        opts.context = {};
      }

      try {
        if (typeof opts.template === 'string') {
          rendered = opts.templateEngine(opts.template, opts.context);
        } else {
          rendered = opts.template(opts.context);
        }
      } catch (e) {
        throw new POE(5, "template error: " + e);
      }

      $(target.DOM.container).html(rendered);
    }


    //   EXTRACTING DOM PARTS
    //   --------------------

    if (!$.isPlainObject(opts.selectors)) {
      opts.selectors = {};
    }

    domParts = $.extractParts(target.DOM.container, opts.selectors);
    $.extend(true, target.DOM, domParts);


    //   HIDE THE CONTAINER IF IT IS NECESSARY
    //   -------------------------------------

    if (opts.hide) {
      $(target.DOM.container).hide();
    }

    return target;
  };

  $.turnToPageObject.configure = function (options) {
    if ($.isPlainObject(options)) {
      $.extend(priv.defaultOptions, options);
    }
  };

  // Expose private things for testing.
  // Don't worry, following lines are not included
  // in minimized production version.
  if (window.PAGE_OBJECT_JS_TEST_MODE) {
    $.turnToPageObject.priv = priv;
  }

}(window, jQuery));


PageObject.js
/*global jQuery: false */
/*jslint browser: true, indent: 2, white: true, nomen: true */

/*
 * PageObject v0.16
 *
 * Copyright 2011-2013, Mykhaylo Gavrylyuk, https://github.com/IronGroove
 * Licensed under the MIT license
 *
 * Date: Thu Aug 01 00:05:21 EEST 2013
 */

(function (window, $) {
  "use strict";

  var priv = {}, POE;

  function PageObjectError(code, message) {
    this.name = "PageObjectError";
    this.code = code < 10 ? "0"+code : code;
    this.message = message;
  }
  function Mediator(){}
  Mediator.prototype = Error.prototype;
  PageObjectError.prototype = new Mediator();
  PageObjectError.prototype.constructor = new PageObjectError;
  PageObjectError.prototype.toString = function () {
    return "PageObjectError #" + this.code + ": " + this.message;
  }

  POE = PageObjectError;

  priv.defaultOptions = {
    container: undefined,
    containerClass: undefined,
    containerElement: 'div',
    templateEngine: undefined,
    template: undefined,
    context: {},
    selectors: {},
    namespace: 'DOM',
    hide: false
  };

  // Use underscore microtemplating function as default if using underscore.
  if (window._ && $.isFunction(window._.template)) {
    priv.defaultOptions.templateEngine = window._.template;
  }

  // Extract jQuery DOM parts from a container using a map of selectors.
  $.extractParts = function (sourceContainer, selectors) {
    var domParts = {};

    if (!$.isElement(sourceContainer)) {
      throw new POE(10, "arguments should be: 1st — HTML DOM container, 2nd — selectors map");
    }

    if (!$.isPlainObject(selectors)) {
      selectors = {};
    }

    $.each(selectors, function (name, selector) {
      var found,
        notSure,
        findMultiple;

      if (!/^[A-z\.]+$/.test(name)) {
        throw new POE(11, "incorrect selector name `" + name + "`");
      }

      if ($.isArray(selector) &&
        selector.length === 2 &&
        $.type(selector[0]) === 'string' &&
        $.isFunction(selector[1])) {

        selector[0] = $.trim(selector[0]);

        if (selector[0].indexOf('?') === 0 || selector[0].indexOf('[]') === 0) {
          throw new POE(12, "[] and ? features are unavailable when selector is an array");
        }

        domParts[name] = {};
        $(sourceContainer).find(selector[0]).each(function () {
          var id = selector[1].call(this, this);
          if (id) {
            if (domParts[name][id]) {
              throw new POE(13, "duplicate identifier `" + id + "` in DOM part namespace `"+ name +"`");
            }
            domParts[name][id] = this;
          }
        });
      }

      else if ($.isPlainObject(selector)) {
        domParts[name] = $.extractParts(sourceContainer, selector);
      }

      else if (typeof selector === 'string') {

        while (true) {
          selector = $.trim(selector);

          if (selector.indexOf('?') === 0) {
            notSure = true;
            selector = $.trim(selector.substring(1));
            continue;
          }

          if (selector.indexOf('[]') === 0) {
            findMultiple = true;
            selector = $.trim(selector.substring(2));
            continue;
          }

          break;
        }

        found = $(sourceContainer).find(selector);
        found = $.makeArray(found);

        if (found.length == 0) {
          if (notSure) {
            found = undefined;
          } else {
            throw new POE(14, "DOM parts weren't found for selector `" + name + "`");
          }
        }

        else if (found.length == 1) {
          if (!findMultiple) {
            found = found[0];
          }
        }

        else if (found.length > 1) {
          if (!findMultiple) {
            throw new POE(15, "multiple DOM parts found for selector `" + name + "`");
          }
        }

        domParts[name] = found;
      }

      else {
        throw new POE(16, "invalid selector value");
      }

    });

    return domParts;
  };

  // http://stackoverflow.com/a/384380/133257
  $.isElement = function (o) {
    return  (
      typeof window.HTMLElement === "object" ? o instanceof window.HTMLElement : // DOM2
      typeof o === "object" && o !== null && o.nodeType === 1 && typeof o.nodeName === "string"
    );
  };

  $.isPageObject = function (po) {
    return po && po.DOM && $.isPlainObject(po.DOM) && $.isElement(po.DOM.container);
  }

  $.turnToPageObject = function (target, options) {
    var opts, rendered, domParts, namespace;


    //   CHECK IF 1ST ARGUMENT IS OBJECT
    //   -------------------------------

    if ($.type(target) !== 'object') {
      throw new POE(1, "invalid target (1st argument should be an object)!");
    }

    if (!$.isPlainObject(options)) {
      options = {};
    }


    opts = $.extend({}, priv.defaultOptions, options);


    if (options.namespace === null) {

      //   DON'T USE NAMESPACE IF THERE'S SUCH NEED
      //   ----------------------------------------

      namespace = target;

    } else {

      //   CREATE DOM NAMESPACE IF THERE'S NO SUCH
      //   ---------------------------------------

      if (!$.isPlainObject(target[opts.namespace])) {
        target[opts.namespace] = {};
      }

      namespace = target[opts.namespace];
    }


    //   PREPARE CONTAINER
    //   -----------------

    if (!$.isElement(namespace.container)) {
      if (opts.container !== undefined) {
        if (!$.isElement(opts.container)) {
          throw new POE(2, "container is invalid!");
        }
        namespace.container = opts.container;
      } else {
        namespace.container = window.document.createElement(opts.containerElement);
      }
    }


    //   SET CONTAINER CLASS WHEN NEEDED
    //   -------------------------------

    if (opts.containerClass) {
      $(namespace.container).addClass(opts.containerClass);
    }


    //   ESTABLISH REFERENCE
    //   -------------------

    $(namespace.container).data('pageObject', target);


    //   RENDER TEMPLATE
    //   ---------------

    if (opts.template !== undefined) {

      if (typeof opts.template === 'string') {
        if (!$.isFunction(opts.templateEngine)) {
          throw new POE(3, "templateEngine not configured");
        }
      } else if (typeof opts.template !== 'function' ) {
        throw new POE(4, "invalid template");
      }


      if (!$.isPlainObject(opts.context)) {
        opts.context = {};
      }

      try {
        if (typeof opts.template === 'string') {
          rendered = opts.templateEngine(opts.template, opts.context);
        } else {
          rendered = opts.template(opts.context);
        }
      } catch (e) {
        throw new POE(5, "template error: " + e);
      }

      $(namespace.container).html(rendered);
    }


    //   EXTRACTING DOM PARTS
    //   --------------------

    if (!$.isPlainObject(opts.selectors)) {
      opts.selectors = {};
    }

    domParts = $.extractParts(namespace.container, opts.selectors);
    $.extend(true, namespace, domParts);


    //   HIDE THE CONTAINER IF IT IS NECESSARY
    //   -------------------------------------

    if (opts.hide) {
      $(namespace.container).hide();
    }

    return target;
  };

  $.turnToPageObject.configure = function (options) {
    if ($.isPlainObject(options)) {
      $.extend(priv.defaultOptions, options);
    }
  };


  $.PageObject = function PageObject(options) {
    if (!(this instanceof PageObject)) {
      return new PageObject(options);
    }

    if (!$.isPlainObject(options)) {
      options = {};
    }

    options.namespace = null;
    $.turnToPageObject(this, options);
  };


  // Expose private things for testing.
  // Don't worry, following lines are not included
  // in minimized production version.
  if (window.PAGE_OBJECT_JS_TEST_MODE) {
    $.turnToPageObject.priv = priv;
  }

}(window, jQuery));
